### PR TITLE
feat(sql): date_trunc() implementation

### LINKED ISSUE #1818: Support DATE_TRUNC
### Is your feature request related to a problem?

When trying to connect to Google Data Studio and graph timeseries, putting Data & Time in the Dimension causes the Postgres connection to throw an error:

```
Failed to execute connection with error: ERROR: unknown function name: DATE_TRUNC(STRING,DATE)
  Position: 13
```

### Describe the solution you'd like.

_No response_

### Describe alternatives you've considered.

_No response_

### Additional context.

_No response_

### PR DESCRIPTION
**Pull request by jerrinot**
closes #1818

It adds support for truncating timestamps. 
Intervals are not supported at the moment. I'm not sure if QuestDB supports the Postgres Interval type at all. Any pointer is much appreciated.

### CODE REVIEW

File: core/src/test/java/io/questdb/griffin/engine/functions/date/DateTruncFunctionFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        assertTimestamp("SELECT DATE_TRUNC('microseconds',  TIMESTAMP '2017-03-17T02:09:30.000000Z') as truncated","2017-03-17T02:09:30.000000Z");
+        assertTimestamp("SELECT DATE_TRUNC('milliseconds',  TIMESTAMP '2017-03-17T02:09:30.111111Z') as truncated","2017-03-17T02:09:30.111000Z");
+        assertTimestamp("SELECT DATE_TRUNC('second',        TIMESTAMP '2017-03-17T02:09:30.111111Z') as truncated","2017-03-17T02:09:30.000000Z");
+        assertTimestamp("SELECT DATE_TRUNC('minute',        TIMESTAMP '2017-03-17T02:09:30.111111Z') as truncated","2017-03-17T02:09:00.000000Z");
+        assertTimestamp("SELECT DATE_TRUNC('hour',          TIMESTAMP '2017-03-17T02:09:30.111111Z') as truncated","2017-03-17T02:00:00.000000Z");
+        assertTimestamp("SELECT DATE_TRUNC('day',           TIMESTAMP '2017-03-17T02:09:30.111111Z') as truncated","2017-03-17T00:00:00.000000Z");
+        assertTimestamp("SELECT DATE_TRUNC('week',          TIMESTAMP '2017-03-17T02:09:30.111111Z') as truncated","2017-03-13T00:00:00.000000Z");
+        assertTimestamp("SELECT DATE_TRUNC('week',          TIMESTAMP '2017-03-13T02:09:30.111111Z') as truncated","2017-03-13T00:00:00.000000Z");
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/griffin/engine/functions/date/DateTruncFunctionFactoryTest.java:47]: Although test coverage is very good, I would expect to see edge case tests. Specifically around newly added date arithmetic.

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by marregui** [comment]
hi @jerrinot, I hope you are well!

I was wondering what the status of the PR is, are you going to continue to work on it?.
Thank you in advance, best regards.

File: core/src/test/java/io/questdb/std/datetime/microtime/TimestampsBruteForceTest.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                throw new AssertionError("More than 12 months in a year, huh? month: " + month);
+        }
+        return temporal.with(MONTH_OF_YEAR, shiftedMonth)
+                .with(DAY_OF_MONTH, 1)
+                .with(MICRO_OF_DAY, 0);
+    });
+
+
+    @Test
+    public void testFloorSS_bruteforce() {
```
Comment (jerrinot) [inline core/src/test/java/io/questdb/std/datetime/microtime/TimestampsBruteForceTest.java:85]: please ginote a left-over name

**Comment by jerrinot** [comment]
@marregui I pushed an idea of randomized brute-force tests, the idea comes from @bluestreak01 
It's not complete yet, just an idea. please let me know what do you think.

**Comment by marregui** [comment]
hola @jerrinot would it make sense to reopen the PR from a branch off questdb/master and take the review there?

brute force meaning exhaustive with the whole domain for the function would take an infinite amount of time.

**Comment by jerrinot** [comment]
@marregui: sure, I'll push my branch to the upstream repo. 

indeed, it's a compromise between test duration and the confidence the test gives. the alternative is to hard-code a few selected "edge-case" timestamps and test just that. this assumes you select the right edge-cases. and if you don't then you may learn about it only when it's too late. that is the idea anyway, my opinion is not overly strong on this.

**Comment by jerrinot** [comment]
closing in favour of https://github.com/questdb/questdb/pull/2017
